Add torch.promote_types function

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26051 Add type promotion support in sparse tensors
* **#26655 Add torch.promote_types function**

Differential Revision: [D17556196](https://our.internmc.facebook.com/intern/diff/D17556196)

nairbv committed:
Add torch.promote_types function

[ghstack-poisoned]

nairbv committed:
Update on "Add torch.promote_types function"




[ghstack-poisoned]

gchanan commented on torch/_torch_docs.py:
```diff
@@ -3976,6 +3976,26 @@ def merge_dicts(*dicts):
+promote_types(type1, type2) -> dtype
+
+Returns the :class:`torch.dtype` with the smallest size and scalar kind to which
+`type1` and `type2` may be safely cast. See type promotion :ref:`documentation <type-promotion-doc>`
```
"safely cast" isn't really defined here and I would argue are casting rules aren't really safe (i.e. int32 + float32) -> float32 can lose precision.  Can you reword this somehow?

nairbv replied:
updated to a more specific explanation.

nairbv committed:
Update on "Add torch.promote_types function"




[ghstack-poisoned]

gchanan commented on torch/_torch_docs.py:
```diff
@@ -3976,6 +3976,27 @@ def merge_dicts(*dicts):
+promote_types(type1, type2) -> dtype
+
+Returns the :class:`torch.dtype` with the smallest size and scalar kind that is
+not smaller or of lower kind than `type1` or `type2`. See type promotion
```
nit: not smaller nor of a lower kind than either

nairbv committed:
Update on "Add torch.promote_types function"


Differential Revision: [D17556196](https://our.internmc.facebook.com/intern/diff/D17556196)

[ghstack-poisoned]

nairbv committed:
Update on "Add torch.promote_types function"


Differential Revision: [D17556196](https://our.internmc.facebook.com/intern/diff/D17556196)

[ghstack-poisoned]

nairbv committed:
Update on "Add torch.promote_types function"


Differential Revision: [D17556196](https://our.internmc.facebook.com/intern/diff/D17556196)

[ghstack-poisoned]

nairbv committed:
Update on "Add torch.promote_types function"


Differential Revision: [D17556196](https://our.internmc.facebook.com/intern/diff/D17556196)

[ghstack-poisoned]

gchanan commented:
does this work with JIT?

nairbv commented:
> does this work with JIT?

Seems like it works as much as scalar types in general work (they're treated as ints):
```
>>> code='''
... def foo():
...     return torch.promote_types(torch.int, torch.double)
... def bar():
...     return torch.double
... '''
>>> cu=torch.jit.CompilationUnit(code)
>>> cu.foo()
7
>>> cu.bar()
7
```
